connectors/spark/src/test/java/io/unitycatalog/spark/auth/AwsCredentialRenewalTest.java

```diff
@@ -278,81 +274,62 @@ private List<Row> sql(String statement, Object... args) {
    */
-  public static class TimeBasedCredentialsGenerator implements CredentialsGenerator {
-
-    @Override
-    public Credentials generate(CredentialContext credentialContext) {
+  public abstract static class TimeBasedCredGenerator<T> {
+    public T generate(CredentialContext ignored) {
       long curTsMillis = testClock().now().toEpochMilli();
       // Align it into the window [starTs, starTs + DEFAULT_INTERVAL_MILLIS].
       long startTsMillis = curTsMillis / DEFAULT_INTERVAL_MILLIS * DEFAULT_INTERVAL_MILLIS;
-      return Credentials.builder()
-          .accessKeyId("accessKeyId" + startTsMillis)
-          .secretAccessKey("secretAccessKey" + startTsMillis)
-          .sessionToken("sessionToken" + startTsMillis)
-          .expiration(Instant.ofEpochMilli(startTsMillis + DEFAULT_INTERVAL_MILLIS))
-          .build();
+      return newTimeBasedCred(startTsMillis);
     }
+
+    protected abstract T newTimeBasedCred(long ts);
   }
 
   /**
-   * A testing S3 filesystem used to verify credential renewal behavior. For each {@code
+   * A testing filesystem used to verify credential renewal behavior. For each {@code
    * checkCredentials()} call, the previous credentials should automatically renew as the 30-second
    * time window advances. The test tracks how many distinct credentials this filesystem receives,
    * which should match the expected number of credential renewals. We use this filesystem to
    * accurately track how many renewal happened.
    */
-  public static class S3CredFileSystem extends CredentialTestFileSystem {
-
+  public abstract static class CredRenewFileSystem<T> extends CredentialTestFileSystem {
     private final Set<Long> verifiedTs = new HashSet<>();
-    private volatile AwsCredentialsProvider lazyProvider;
+    private volatile T lazyProvider;
 
     @Override
     protected void checkCredentials(Path f) {
       String host = f.toUri().getHost();
 
       if (credentialCheckEnabled && BUCKET_NAME.equals(host)) {
-        AwsCredentialsProvider provider = accessProvider();
+        T provider = accessProvider();
         assertThat(provider).isNotNull();
 
-        AwsSessionCredentials cred = (AwsSessionCredentials) provider.resolveCredentials();
-        long ts =
-            testClock().now().toEpochMilli() / DEFAULT_INTERVAL_MILLIS * DEFAULT_INTERVAL_MILLIS;
-        assertThat(cred.accessKeyId()).isEqualTo("accessKeyId" + ts);
-        assertThat(cred.secretAccessKey()).isEqualTo("secretAccessKey" + ts);
-        assertThat(cred.sessionToken()).isEqualTo("sessionToken" + ts);
+        long curTs = testClock().now().toEpochMilli();
+        long windowStartTs = curTs / DEFAULT_INTERVAL_MILLIS * DEFAULT_INTERVAL_MILLIS;
+        assertCredentials(provider, windowStartTs);
 
-        verifiedTs.add(ts);
+        verifiedTs.add(windowStartTs);
       }
     }
 
-    public int renewalCount() {
-      return verifiedTs.size() - 1;
-    }
-
-    @Override
-    protected String scheme() {
```

Comment:
This will be overrided in the cloud vendor `CredRenewFileSystem`, such as `S3CredFileSystem`, and `AbfsCredFileSystem`.